File: ff4j-spring-services/src/main/java/org/ff4j/services/FeatureServices.java
package org.ff4j.services;

import org.ff4j.FF4j;
import org.ff4j.core.Feature;
import org.ff4j.exception.FeatureAccessException;
import org.ff4j.services.domain.FeatureApiBean;
import org.ff4j.services.domain.FlippingStrategyApiBean;
import org.ff4j.services.domain.PropertyApiBean;
import org.ff4j.services.exceptions.FlippingStrategyBadRequestException;
import org.ff4j.services.exceptions.PropertiesBadRequestException;
import org.ff4j.services.model.FeatureActions;
import org.ff4j.services.validator.FeatureValidator;
import org.ff4j.utils.MappingUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.Map;

/**
 * Created by Paul
 *
 * @author <a href="mailto:paul58914080@gmail.com">Paul Williams</a>
 */
@Service
public class FeatureServices {
    @Autowired
    private FF4j ff4j;
    @Autowired
    private FeatureValidator featureValidator;

    public FeatureApiBean getFeature(String featureUID) {
        featureValidator.assertFeatureExists(featureUID);
        return new FeatureApiBean(ff4j.getFeatureStore().read(featureUID));
    }

    public FeatureActions createOrUpdateFeature(String featureUID, FeatureApiBean featureApiBean) {
        featureValidator.assertFeatureUIDIsNotBlank(featureApiBean.getUid());
        featureValidator.assertFeatureIdsMatch(featureUID, featureApiBean.getUid());
        Feature feature = new Feature(featureUID);
        feature.setDescription(featureApiBean.getDescription());
        feature.setEnable(featureApiBean.getEnable());
        feature.setGroup(featureApiBean.getGroup());
        feature.setPermissions(new HashSet<>(featureApiBean.getPermissions()));
        initFlippingStrategy(featureApiBean, feature);
        initProperties(featureApiBean, feature);
        if (ff4j.getFeatureStore().exist(featureUID)) {
            ff4j.getFeatureStore().update(feature);
            return FeatureActions.UPDATED;
        } else {
            ff4j.getFeatureStore().create(feature);
            return FeatureActions.CREATED;
        }
    }

    private void initProperties(FeatureApiBean featureApiBean, Feature feature) {
        Map<String, PropertyApiBean> mapProperties = featureApiBean.getCustomProperties();
        if (mapProperties != null) {
            try {
                for (PropertyApiBean propertyBean : mapProperties.values()) {
                    feature.addProperty(propertyBean.asProperty());
                }
            } catch (IllegalArgumentException exception) {
                throw new PropertiesBadRequestException(exception);
            }
        }
    }

    private void initFlippingStrategy(FeatureApiBean featureApiBean, Feature feature) {
        FlippingStrategyApiBean flipApiBean = featureApiBean.getFlippingStrategy();
        if (null != flipApiBean) {
            try {
                Map<String, String> initParams = flipApiBean.getInitParams();
                feature.setFlippingStrategy(MappingUtil.instanceFlippingStrategy(featureApiBean.getUid(), flipApiBean.getType(), initParams));
            } catch (FeatureAccessException exception) {
                throw new FlippingStrategyBadRequestException(exception);
            }
        }
    }

    public void deleteFeature(String featureUID) {
        featureValidator.assertFeatureExists(featureUID);
        ff4j.getFeatureStore().delete(featureUID);
    }

    public void enableFeature(String featureUID) {
        featureValidator.assertFeatureExists(featureUID);
        ff4j.getFeatureStore().enable(featureUID);
    }

    public void disableFeature(String featureUID) {
        featureValidator.assertFeatureExists(featureUID);
        ff4j.getFeatureStore().disable(featureUID);
    }

    public void grantRoleToFeature(String featureUID, String role) {
        featureValidator.assertFeatureExists(featureUID);
        featureValidator.assertRoleDoesNotExist(featureUID, role);
        ff4j.getFeatureStore().grantRoleOnFeature(featureUID, role);
    }

    public void removeRoleFromFeature(String featureUID, String role) {
        featureValidator.assertFeatureExists(featureUID);
        featureValidator.assertRoleExist(featureUID, role);
        ff4j.getFeatureStore().removeRoleFromFeature(featureUID, role);
    }

    public void addGroupToFeature(String featureUID, String groupName) {
        featureValidator.assertFeatureExists(featureUID);
        featureValidator.assertGroupDoesNotExist(groupName);
        ff4j.getFeatureStore().addToGroup(featureUID, groupName);
    }

    public void removeGroupFromFeature(String featureUID, String groupName) {
        featureValidator.assertFeatureExists(featureUID);
        featureValidator.assertGroupExist(groupName);
        ff4j.getFeatureStore().removeFromGroup(featureUID, groupName);
    }
}

File: ff4j-spring-services/src/main/kotlin/org/ff4j/services/FeatureServices.kt
/*-
 * #%L
 * ff4j-spring-services
 * %%
 * Copyright (C) 2013 - 2022 FF4J
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package org.ff4j.services

import org.ff4j.FF4j
import org.ff4j.services.domain.FeatureApiBean
import org.ff4j.services.model.FeatureActions
import org.ff4j.services.validator.FeatureValidator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

/**
 * Created by Paul
 *
 * @author <a href="mailto:paul58914080@gmail.com">Paul Williams</a>
 */
@Service
class FeatureServices(@Autowired val fF4j: FF4j, @Autowired val featureValidator: FeatureValidator) {
  fun getFeature(featureUID: String): Mono<FeatureApiBean> {
    featureValidator.assertFeatureExists(featureUID)
    return Mono.just(FeatureApiBean(fF4j.featureStore.read(featureUID)))
  }

  fun createOrUpdateFeature(featureUID: String, featureApiBean: FeatureApiBean): Mono<FeatureActions> {
    featureValidator.assertFeatureUIDIsNotBlank(featureApiBean.uid)
    featureValidator.assertFeatureIdsMatch(featureUID, featureApiBean.uid)
    return if (fF4j.featureStore.exist(featureUID)) {
      fF4j.featureStore.update(featureApiBean.toFeature())
      Mono.just(FeatureActions.UPDATED)
    } else {
      fF4j.featureStore.create(featureApiBean.toFeature())
      Mono.just(FeatureActions.CREATED)
    }
  }

  fun deleteFeature(featureUID: String) {
    featureValidator.assertFeatureExists(featureUID)
    fF4j.featureStore.delete(featureUID)
  }

  fun disableFeature(featureUID: String) {
    featureValidator.assertFeatureExists(featureUID)
    fF4j.featureStore.disable(featureUID)
  }

  fun enableFeature(featureUID: String) {
    featureValidator.assertFeatureExists(featureUID)
    fF4j.featureStore.enable(featureUID)
  }

  fun grantRoleToFeature(featureUID: String, roleName: String) {
    featureValidator.assertFeatureExists(featureUID)
    featureValidator.assertRoleDoesNotExist(featureUID, roleName)
    fF4j.featureStore.grantRoleOnFeature(featureUID, roleName)
  }

  fun removeRoleFromFeature(featureUID: String, roleName: String) {
    featureValidator.assertFeatureExists(featureUID)
    featureValidator.assertRoleExist(featureUID, roleName)
    fF4j.featureStore.removeRoleFromFeature(featureUID, roleName)
  }

  fun addGroupToFeature(featureUID: String, groupName: String) {
    featureValidator.assertFeatureExists(featureUID)
    featureValidator.assertGroupDoesNotExist(groupName)
    fF4j.featureStore.addToGroup(featureUID, groupName)
  }

  fun removeGroupFromFeature(featureUID: String, groupName: String) {
    featureValidator.assertFeatureExists(featureUID)
    featureValidator.assertGroupExist(groupName)
    fF4j.featureStore.removeFromGroup(featureUID, groupName)
  }
}
